distill/sessions/session.py

```diff
@@ -0,0 +1,270 @@
+        A user session is defined by: unique session ID, user ID, and separated by idle time
+        that exceeds the specified inactive_interval (in seconds). By default, the interval is 60 seconds.
+        This set is further separated by the windows tab in which the user activities occurred
+        :param log: Userale log in the form of dictionary
+        :param inactive_interval_s: Threshold of inactivity (no logged activity) in seconds
+        :param url_re: Regular expression to filter the log
+        :param group_by_type: either group by tab, URL, browser (None)
+        :return: A dictionary that represent sets of user sessions
+        """
+        data_by_users = {}
+        data_by_users = Session.__group_by_user(log)
+
+        chunk_data_by_users_sessions = {}
+        for user in data_by_users:
+            user_session_sets = {}
+            # Sort the logs associated by each users so we can create sets accordingly
+            sorted_data = sorted(data_by_users[user], key=lambda item: item.get('clientTime', item.get('endTime')))
+            chunked_group = {}
+            # Separate by browser tab
+            if (group_by_type == "tab"):
+                chunked_group = Session.__chunk_by_tabId(sorted_data)
+                for g_id in chunked_group:
+                    # For each set, detect if there is an idle time between the logs that exceed X seconds
+                    user_session_sets[g_id] = Session.__chunk_by_idle_time(chunked_group[g_id], inactive_interval_s)
+                chunk_data_by_users_sessions[user] = user_session_sets
+            # Separate by domain application
+            elif (group_by_type == "domain"):
+                # Do something
+                chunked_group = Session.__chunk_by_domain(sorted_data, url_re)
+                for g_id in chunked_group:
+                    # For each set, detect if there is an idle time between the logs that exceed X seconds
+                    user_session_sets[g_id] = Session.__chunk_by_idle_time(chunked_group[g_id], inactive_interval_s)
+                chunk_data_by_users_sessions[user] = user_session_sets
+            else:
+                # For each set, detect if there is an idle time between the logs that exceed X seconds
+                chunk_data_by_users_sessions[user] = Session.__chunk_by_idle_time(sorted_data, inactive_interval_s)
+
+        # Flatten the structure into a collection of sessions
+        flattened_results = Session.__flatten_dict(chunk_data_by_users_sessions)
+        parsed_sessions = []
+        for result in flattened_results:
+            parsed_sessions.append(Session({result: flattened_results[result]}))
+        sessions_set = Sessions(parsed_sessions)
+        if (group_by_type == "tab"):
+            sessions_set.sessions_type = Sessions_Type.TAB
+        elif (group_by_type == "domain"):
+            sessions_set.sessions_type = Sessions_Type.DOMAIN
+        return sessions_set
```

Comment:
So this class strikes me as a bit weird. In other words. Even though your tests pass, it's not a very pythonic way of implement a data object in python.

 There's have a mixture of OOP and functional paradigms going on here.
 
 For instance, this method is a staticmethod (does not operate on the class or itself/instance of class). You are using this method to build new instances of a `Session**s**` object but doing it as a method on the `Session`. That's misleading.
 
 The parameters for this function are properties I would expect to be set when instantiating a new Sessions object. I could then either pass the logs while instantiating the object, add them later via a method, or both.